Source/Tests/LinuxSyscalls/SyscallsVMATracking.cpp

```diff
@@ -0,0 +1,367 @@
+  // Top is the address after the end
+  auto CurrentIter = VMAs.lower_bound(Top);
+
+  // Iterate backwards all mappings
+  while (CurrentIter != VMAs.begin()) {
+    CurrentIter--;
+
+    const auto Current = &CurrentIter->second;
+    const auto MapBase = Current->Base;
+    const auto MapTop = MapBase + Current->Length;
+    const auto OffsetDiff = Current->Offset - MapBase;
+
+    if (MapTop <= Base) {
+      // Mapping ends before the Range start, exit
+      break;
+    } else {
+      const bool HasFirstPart = MapBase < Base;
+      const bool HasTrailingPart = MapTop > Top;
+
+      // (1) HasFirstPart, !HasTrailingPart -> trim
+      // (2) HasFirstPart, HasTrailingPart -> trim, insert trailing, list add after first part
+      // (3) !HasFirstPart, !HasTrailing part -> list remove, erase
+      // (4) !HasFirstPart, HasTrailing part -> insert trailing, list replace first part, erase
+
+      if (HasFirstPart) {
+        // Handle trim for (1) & (2)
+        Current->Length = Base - MapBase;
+      } else if (!HasTrailingPart) {
+        // Handle all of (3)
+        // Mapping is included or equal to Range, delete
+
+        // If linked to a Mapped Resource, remove from linked list and possibly delete the Mapped Resource
+        if (Current->Resource) {
+          if (ListRemove(Current) && Current->Resource != PreservedMappedResource) {
+            if (Current->Resource->AOTIRCacheEntry) {
+              FEXCore::Context::UnloadAOTIRCacheEntry(CTX, Current->Resource->AOTIRCacheEntry);
+            }
+            MappedResources.erase(Current->Resource->Iterator);
+          }
+        }
+
+        // returns next element, so -- is safe at loop
+        CurrentIter = VMAs.erase(CurrentIter);
+        continue; // we're done
+      }
+
+      const bool ReplaceAndErase = !HasFirstPart;
+
+      if (HasTrailingPart) {
+        // Handle insert of (2), (4)
+
+        // insert trailing part, link it after Mapping
+        auto NewOffset = OffsetDiff + Top;
+        auto NewLength = MapTop - Top;
+
+        auto Inserted = VMAs.emplace(
+            Top, VMAEntry{Current->Resource, ReplaceAndErase ?  Current->ResourcePrevVMA : Current, Current->ResourceNextVMA, Top, NewOffset, NewLength, Current->Options});
+        LOGMAN_THROW_A_FMT(Inserted.second, "VMA tracking error");
+        auto TrailingPart = &Inserted.first->second;
+        if (Current->Resource) {
+          if (ReplaceAndErase) {
+            // Handle list replace of (4)
+            ListReplace(Current, TrailingPart);
+          } else {
+            // Handle list insert (2)
+            ListInsert(Current, TrailingPart);
+          }
+        }
+      }
+
+      if (ReplaceAndErase) {
+        // Handle erase of (4)
+        // returns next element, so -- is safe at loop
+        CurrentIter = VMAs.erase(CurrentIter);
+      }
+    }
+  }
+}
+
+// Change flags of mappings in a range and split the mappings if needed
```

Comment:
(see docstring comment above).
Those aren't actually meant to be documentation for the users, but for the implementer, that's why they are there.

Reply:
Because this is a private api, only consumed within this file, and this feature is also easy to deduce as trackMmap/Munmap that are the only users of this also behave this way.